From 9d104bfa80010e064f00eb5ce0e2f3ccf522761e Mon Sep 17 00:00:00 2001
From: Jan Remes <jan.remes@codasip.com>
Date: Mon, 6 Dec 2021 18:17:16 +0100
Subject: [PATCH 01/12] board: codasip: add support for Codasip A70X cores on
 FPGA platform

The defconfig and DTS will be kept in buildroot external
---
 arch/riscv/Kconfig              |  4 ++++
 arch/riscv/dts/Makefile         |  1 +
 arch/riscv/dts/codasip-a70x.dts |  9 +++++++++
 board/codasip/a70x/Kconfig      | 18 +++++++++++++++++
 board/codasip/a70x/Makefile     |  1 +
 board/codasip/a70x/platform.c   | 26 ++++++++++++++++++++++++
 include/configs/codasip-a70x.h  | 36 +++++++++++++++++++++++++++++++++
 7 files changed, 95 insertions(+)
 create mode 100644 arch/riscv/dts/codasip-a70x.dts
 create mode 100644 board/codasip/a70x/Kconfig
 create mode 100644 board/codasip/a70x/Makefile
 create mode 100644 board/codasip/a70x/platform.c
 create mode 100644 include/configs/codasip-a70x.h

diff --git a/arch/riscv/Kconfig b/arch/riscv/Kconfig
index 1c62c2345b..b9bb4d1934 100644
--- a/arch/riscv/Kconfig
+++ b/arch/riscv/Kconfig
@@ -39,6 +39,9 @@ config TARGET_SIPEED_MAIX
 config TARGET_OPENPITON_RISCV64
 	bool "Support RISC-V cores on OpenPiton SoC"
 
+config TARGET_CODASIP_A70X
+	bool "Support Codasip A70X FPGA Platform"
+
 endchoice
 
 config SYS_ICACHE_OFF
@@ -83,6 +86,7 @@ source "board/thead/th1520_lpi4a/Kconfig"
 source "board/openpiton/riscv64/Kconfig"
 source "board/sipeed/maix/Kconfig"
 source "board/starfive/visionfive2/Kconfig"
+source "board/codasip/a70x/Kconfig"
 
 # platform-specific options below
 source "arch/riscv/cpu/andesv5/Kconfig"
diff --git a/arch/riscv/dts/Makefile b/arch/riscv/dts/Makefile
index f1525cb668..9e055d2ce1 100644
--- a/arch/riscv/dts/Makefile
+++ b/arch/riscv/dts/Makefile
@@ -9,6 +9,7 @@ dtb-$(CONFIG_TARGET_SIFIVE_UNMATCHED) += hifive-unmatched-a00.dtb
 dtb-$(CONFIG_TARGET_SIPEED_MAIX) += k210-maix-bit.dtb
 dtb-$(CONFIG_TARGET_STARFIVE_VISIONFIVE2) += jh7110-starfive-visionfive-2.dtb
 dtb-$(CONFIG_TARGET_TH1520_LPI4A) += th1520-lichee-pi-4a.dtb
+dtb-$(CONFIG_TARGET_CODASIP_A70X) += codasip-a70x.dtb
 include $(srctree)/scripts/Makefile.dts
 
 targets += $(dtb-y)
diff --git a/arch/riscv/dts/codasip-a70x.dts b/arch/riscv/dts/codasip-a70x.dts
new file mode 100644
index 0000000000..ad184839ed
--- /dev/null
+++ b/arch/riscv/dts/codasip-a70x.dts
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: GPL-2.0+ OR MIT
+/*
+ * Empty device tree
+ */
+
+/dts-v1/;
+
+/ {
+};
diff --git a/board/codasip/a70x/Kconfig b/board/codasip/a70x/Kconfig
new file mode 100644
index 0000000000..1276a76004
--- /dev/null
+++ b/board/codasip/a70x/Kconfig
@@ -0,0 +1,18 @@
+if TARGET_CODASIP_A70X
+
+config SYS_BOARD
+	default "a70x"
+
+config SYS_VENDOR
+	default "codasip"
+
+config SYS_CPU
+	default "generic"
+
+config SYS_CONFIG_NAME
+	default "codasip-a70x"
+
+config TEXT_BASE
+	default 0x80000000 if !RISCV_SMODE
+	default 0x80200000 if RISCV_SMODE
+endif
diff --git a/board/codasip/a70x/Makefile b/board/codasip/a70x/Makefile
new file mode 100644
index 0000000000..987b39e5e9
--- /dev/null
+++ b/board/codasip/a70x/Makefile
@@ -0,0 +1 @@
+obj-y := platform.o
diff --git a/board/codasip/a70x/platform.c b/board/codasip/a70x/platform.c
new file mode 100644
index 0000000000..6a8913588d
--- /dev/null
+++ b/board/codasip/a70x/platform.c
@@ -0,0 +1,26 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+
+#include <cpu_func.h>
+#include <dm.h>
+#include <asm/sections.h>
+
+int misc_init_r(void)
+{
+	return 0;
+}
+
+void *board_fdt_blob_setup(int *err)
+{
+	*err = 0;
+	if (IS_ENABLED(CONFIG_OF_SEPARATE) || IS_ENABLED(CONFIG_OF_BOARD)) {
+		if (gd->arch.firmware_fdt_addr)
+			return (ulong *)(uintptr_t)gd->arch.firmware_fdt_addr;
+	}
+
+	return (ulong *)&_end;
+}
+
+int board_init(void)
+{
+	return 0;
+}
diff --git a/include/configs/codasip-a70x.h b/include/configs/codasip-a70x.h
new file mode 100644
index 0000000000..f5a5fede74
--- /dev/null
+++ b/include/configs/codasip-a70x.h
@@ -0,0 +1,36 @@
+#ifndef CODASIP_A70X_H
+#define CODASIP_A70X_H
+
+#include <linux/sizes.h>
+
+#define CONFIG_SYS_SDRAM_BASE		0x80000000
+#define CONFIG_EXTRA_ENV_SETTINGS \
+	"fdt_high=0xffffffffffffffff\0" \
+	"mmcdev=0\0" \
+	"bootenvfile=uEnv.txt\0" \
+	"linuxfile=Image\0" \
+	"linuxaddr=0x81000000\0" \
+	"dtbaddr=0x86000000\0" \
+	"envboot=fatload mmc 0 ${loadaddr} ${bootenvfile} && " \
+		"env import -t ${loadaddr} ${filesize};\0" \
+	"downloaddtb=${netcmd} ${dtbaddr} ${serverip}:${dtbfile};\0" \
+	"downloadlinux=${netcmd} ${linuxaddr} ${serverip}:${linuxfile};\0" \
+	"netboot=if test -z \"${serverip}\"; then " \
+			"echo No TFTP server IP address specified, set the variable serverip.; " \
+		"else " \
+			"if test -n ${ipaddr}; then " \
+				"setenv netcmd tftpboot; " \
+			"else " \
+				"setenv netcmd dhcp; " \
+			"fi; " \
+			"if test -n ${dtbfile}; then " \
+				"run downloaddtb && " \
+				"run downloadlinux && " \
+				"booti ${linuxaddr} - ${dtbaddr}; " \
+			"else " \
+				"run downloadlinux && " \
+				"boot; " \
+			"fi; " \
+		"fi;\0" \
+
+#endif /* CODASIP_A70X_H */
-- 
2.39.3

